src/main/java/reactor/netty/http/MicrometerHttpMetricsRecorder.java

```diff
@@ -21,82 +21,126 @@
+
+	final Counter.Builder errorsBuilder;
+	final ConcurrentMap<MeterKey, Counter> errorsCache = PlatformDependent.newConcurrentHashMap();
+
+	protected MicrometerHttpMetricsRecorder(String name, String protocol) {
+		super(name, protocol);
+		this.dataReceivedTimeBuilder =
+				Timer.builder(name + DATA_RECEIVED_TIME)
+				     .description("Time that is spent in consuming incoming data");
+
+		this.dataSentTimeBuilder =
+				Timer.builder(name + DATA_SENT_TIME)
+				     .description("Time that is spent in sending outgoing data");
 
-	protected MicrometerHttpMetricsRecorder(String name, @Nullable String remoteAddress, String protocol) {
-		super(name, remoteAddress, protocol);
-
-		dataReceivedTimeBuilder = Timer.builder(name + DATA_RECEIVED_TIME)
-		                               .description("Time that is spent in consuming incoming data");
-		dataSentTimeBuilder = Timer.builder(name + DATA_SENT_TIME)
-		                           .description("Time that is spent in sending outgoing data");
-		responseTimeBuilder = Timer.builder(name + RESPONSE_TIME)
-		                           .description("Total time for the request/response");
-
-		registry.config()
-		        .meterFilter(maxUriTagsMeterFilter(name + DATA_RECEIVED_TIME))
-		        .meterFilter(maxUriTagsMeterFilter(name + DATA_SENT_TIME))
-		        .meterFilter(maxUriTagsMeterFilter(name + RESPONSE_TIME))
-		        .meterFilter(maxUriTagsMeterFilter(name + DATA_RECEIVED))
-		        .meterFilter(maxUriTagsMeterFilter(name + DATA_SENT))
-		        .meterFilter(maxUriTagsMeterFilter(name + ERRORS));
+		this.responseTimeBuilder =
+				Timer.builder(name + RESPONSE_TIME)
+				     .description("Total time for the request/response");
+
+		this.dataReceivedBuilder =
+				DistributionSummary.builder(name + DATA_RECEIVED)
+				                   .baseUnit("bytes")
+				                   .description("Amount of the data that is received, in bytes");
+
+		this.dataSentBuilder =
+				DistributionSummary.builder(name + DATA_SENT)
+				                   .baseUnit("bytes")
+				                   .description("Amount of the data that is sent, in bytes");
+
+		this.errorsBuilder =
+				Counter.builder(name + ERRORS)
+				       .description("Number of the errors that are occurred");
+
+		registry.config().meterFilter(maxUriTagsMeterFilter(name));
 	}
 
 	@Override
 	public void recordDataReceived(SocketAddress remoteAddress, String uri, long bytes) {
-		DistributionSummary.builder(name + DATA_RECEIVED)
-		                   .baseUnit("bytes")
-		                   .description("Amount of the data that is received, in bytes")
-		                   .tags(REMOTE_ADDRESS, Metrics.formatSocketAddress(remoteAddress), URI, uri)
-		                   .register(registry)
-		                   .record(bytes);
+		String address = Metrics.formatSocketAddress(remoteAddress);
+		MeterKey meterKey = MeterKey.builder().uri(uri).remoteAddress(address).build();
```

Comment:
Since we're trying to optimize this particularly hot path, it seems like a direct instantiation of a MeterKey object would be better than using a temporary builder

Reply:
agree